[ #2487 ] consistency checking of options

fredrikNordvallForsberg opened the pull request:
This is addressing #2487. Doing a pull request, since there are some choices to perhaps discuss:

* I've removed `--safe` and `--only-scope-checking` being incompatible, since it is otherwise impossible to use `--only-scope-checking` on modules marked as safe (e.g. the QuickLatex tests importing builtin modules). Am I missing some reason why this combination was not allowed? It seems harmless to me.

* I've turned on both `--guardedness` and `--sized-types` by default, but made `--safe` turn them both off.

* I've made the following options "coinfective": they are only allowed to import modules also using the option in question. I think this is more intuitive than making the dual options infective (if they exist).
   - ``--safe``
   - ``--without-K``
   - ``--no-universe-polymorphism``
   - ``--no-sized-types``
   - ``--no-guardedness``

* I've made the following options "infective": they are only allowed to be imported by modules also using the option in question.
  - ``--cubical``
  - ``--prop``

  **Question:** Is there any reason to have ``--cubical`` infective? (For ``--prop``, see #3333.) If possible, 
maybe it shouldn't be, in order to not break current code?

fredrikNordvallForsberg committed:
[ #2487 ] consistency checking of options

jespercockx commented:
LGTM, thanks for the hard work! 

I think `--safe` and `--only-scope-checking` should be incompatible, since `--only-scope-checking` disables almost **all** checks so it's definitely not safe. However, it should of course be possible to import a `--safe` file from a file using `--only-scope-checking`.

The lists of infective and coinfective options make sense to me. I think there are good reasons for making `--cubical` infective, since anything you import from a `--cubical` file may rely on the flag being there for its computation.

nad commented:
The `--only-scope-checking` option was introduced to make certain uses of the LaTeX backend faster. It can not be used as a pragma, and only applies to the top-level module. When this option is used no interface files are written to disk.

fredrikNordvallForsberg commented:
> I think --safe and --only-scope-checking should be incompatible, since --only-scope-checking disables almost all checks so it's definitely not safe. However, it should of course be possible to import a --safe file from a file using --only-scope-checking.

But is it not clear that ``--only-scope-checking`` isn't doing any "real" checking? (In particular, no interface files are generated.)

The option ``--only-scope-checking`` cannot be set in a pragma (that wouldn't make any sense), so the only way to turn it on is globally as a command-line option, which means that any imported files with a pragma ``--safe`` will be checked with both ``--safe`` and ``--only-scope-checking``.

nad commented:
Some questions (related to #2783):

* Why are some of the builtin modules related to `--cubical` not marked as safe?

* Should `Agda.Builtin.Reflection` be marked as safe? My guess would be that it is safe, because the `TC` definitions do not compute (unless they are used by macros), and all definitions can be implemented if `TC` is instantiated to `λ _ → ⊤`.

* Similarly, should `Agda.Builtin.IO` be marked as safe?

I'd like to see some test which uses command line options in combination with pragma options. For instance something like the following:

* Type check the following module using the command-line option `--safe`:
  ```agda
  module A where
  ```

* Type check the following module using no command-line options:
  ```agda
  {-# OPTIONS --safe #-}
  module B where
  import A
  ```

* Type check `A` again.

* Type check `B` again. This last step should fail.

I also think it makes sense to document what options are infective or coinfective (along with these terms) in the user manual, as well as the set of unsafe options.

nad commented:
Did you address the part about rechecking code when certain options (like `--termination-depth`) change?

nad commented:
> which means that any imported files with a pragma ``--safe`` will be checked with both ``--safe`` and ``--only-scope-checking``.

This option only has an effect for the top-level module, so it could be omitted when other modules are checked. However, I think it should be possible to use `--only-scope-checking` for a top-level module which is marked as safe.

fredrikNordvallForsberg commented:
> Why are some of the builtin modules related to --cubical not marked as safe?

``Agda.Builtin.Cubical.Path``, ``Agda.Builtin.Cubical.Sub`` and ``Agda.Builtin.Cubical.Id`` uses postulates, and ``Agda.Builtin.Cubical.Glue`` imports ``...Path`` and ``...Sub``.

> Should Agda.Builtin.Reflection be marked as safe?

> Similarly, should Agda.Builtin.IO be marked as safe?

Also both of these currently use postulates. But perhaps you rather mean "Should we make these safe a la #2783?" -- perhaps this is better discussed in that issue? (However at least for the cubical builtins, this should probably be fixed before releasing.)

> I'd like to see some test which uses command line options in combination with pragma options.

Yes, this is a good idea.

> - Type check the following module using the command-line option --safe:
> ```module A where```
> - Type check the following module using no command-line options:
> ```
> {-# OPTIONS --safe #-}
> module B where
> import A
> ```
> - Type check A again.
> - Type check B again. This last step should fail.

This last step doesn't fail with my implementation, because when type checking A again without ``--safe``, the existing (safe) interface is found to be sufficient for the current (not safe) options. Do you think this is a problem? You cannot use this to trick e.g. ``agda --safe`` into accepting an unsafe file.

> I also think it makes sense to document what options are infective or coinfective (along with these terms) in the user manual, as well as the set of unsafe options.

Yes. I'm not particularly fond of the name "coinfective" in user-facing situations, does anyone have a better suggestion?

> Did you address the part about rechecking code when certain options (like --termination-depth) change?

No, I did not. I'm not sure how easy this would be to achieve.

> This option only has an effect for the top-level module, so it could be omitted when other modules are checked. However, I think it should be possible to use --only-scope-checking for a top-level module which is marked as safe.

I agree. Also because omitting it for other modules would mean that if A is marked safe, then only-scope-checking A directly would be an error, but only-scope-checking a module B which only imports A would succeed, which seems counter-intuitive to me.

nad commented:
> perhaps this is better discussed in that issue?

OK, I'm moving the discussion there.

> This last step doesn't fail with my implementation, because when type checking A again without ``--safe``, the existing (safe) interface is found to be sufficient for the current (not safe) options.

OK. That's fine, I guess.

> > Did you address the part about rechecking code when certain options (like --termination-depth) change?
>
> No, I did not. I'm not sure how easy this would be to achieve.

A simple way to fix this issue might be to disallow giving these options on the command line. At the moment you can do the following:
```
$ cat Test.agda
open import Agda.Builtin.Bool

f : Bool → Bool
f true = true
f _    = false
$ (agda -Werror --exact-split Test.agda > /dev/null) && echo success
$ (agda Test.agda > /dev/null) && echo success
success
$ (agda -Werror --exact-split Test.agda > /dev/null) && echo success
success
```

However, perhaps it should be possible to give (some of) these options on the command line, for instance to verify that some development does not use non-exact splits, or that it does not use non-standard termination depths.

gallais commented:
Shouldn't `Agda/Builtin/Equality/Erase.agda` be marked as safe too?

fredrikNordvallForsberg commented:
> Shouldn't `Agda/Builtin/Equality/Erase.agda` be marked as safe too?

Indeed, that was the conclusion of #3315. Fixing.

fredrikNordvallForsberg committed:
[ fix #2783 ] Most builtin modules may use postulates and still be safe

Reverts part of commit e134cce as we can now postulate String,
Char, Float, Word64 and the coinduction builtins again.

gallais commented on src/full/Agda/Interaction/Highlighting/Generate.hs:
```diff
@@ -165,7 +165,7 @@ generateAndPrintSyntaxInfo
 generateAndPrintSyntaxInfo decl hlLevel updateState = do
-  file <- fromMaybe __IMPOSSIBLE__ <$> asksTC envCurrentPath
+  file <- getCurrentPath
 
```
Should `getCurrentPath` take an `__IMPOSSIBLE__` as an argument in order to preserve the
quality of the location information in the internal error?

fredrikNordvallForsberg replied:
It's only used in a few places, so probably not too hard to track down. And except for in the deprecated interactive mode, I think this will always be a `Just` anyway... For now, I prefer the cleaner code, but happy to change this if an `__IMPOSSIBLE__` actually shows up!

gallais replied:
Fair enough!

gallais commented on src/full/Agda/Interaction/Imports.hs:
```diff
@@ -410,12 +427,26 @@ getInterface' x isMain msi = do
+      return (i, wt')
+        where
+          checkOptionsCompatible :: OptionsUsed -> OptionsUsed -> ModuleName -> TCM Bool
+          checkOptionsCompatible current imported importedModule = flip execStateT True $ do
```
Should we lift this to the toplevel (maybe near the definition of the `OptionsUsed` type?)

fredrikNordvallForsberg replied:
Top-level is a good idea, but unfortunately it seems hard to move it to `TypeChecking.Options` without introducing import cycles...

gallais commented on src/full/Agda/Interaction/Imports.hs:
```diff
@@ -698,6 +731,15 @@ writeInterface file i = do
 
+concreteOptionsToOptionPragmas :: [C.Pragma] -> TCM [OptionsPragma]
+concreteOptionsToOptionPragmas p = do
+  pragmas <- concat <$> concreteToAbstract_ p
```
Should we move this next to `OptionsPragma`?
(i.e. in Agda/Interaction/Options.hs)

gallais commented on src/full/Agda/Interaction/Options/Lenses.hs:
```diff
@@ -131,6 +133,14 @@ modifySafeMode = modifyTC . mapSafeMode
+        [ libdir </> "prim" </> "Agda" </> "Builtin" </> "TrustMe.agda"
+        , libdir </> "prim" </> "Agda" </> "Builtin" </> "Equality" </> "Rewrite.agda"
+        ]
+  return $ (prefixPath libdir file) && not (elem file unsafeBuiltins)
```
It would be a bit more tedious but also a bit more bug-proof to list the
files for which the postulates are considered safe rather than assuming
all `prim` files are safe unless explicitly excluded.

fredrikNordvallForsberg replied:
Yes, probably a good idea. Maybe it would also be good to add a test that all builtin files are either safe or unsafe wrt postulates.

gallais commented on src/full/Agda/TypeChecking/Monad/Base.hs:
```diff
@@ -771,6 +772,57 @@ type DecodedModules = Map C.TopLevelModuleName Interface
 data ForeignCode = ForeignCode Range String
   deriving Show
 
+data OptionKeys = SafeOption
```
Should all of these new additions go into `Interaction.Options`?

gallais reviewed (approved):
Mostly nitpicking about how the code is organized.
Really excited by this feature!

fredrikNordvallForsberg committed:
[ #2487 ] restarting if options change, update user-manual

Also adds test cases

gallais committed:
[ #2487 ] Use Utils.getAgdaFilesInDir instead of using listDirectory

gallais commented:
Looks like using `listDirectory` means we would have to drop support for 7.10.3.
I suppose the sensible alternative is to use `System.FilePath.Find` like we do
in [test/Utils](https://github.com/agda/agda/blob/323f58f9b8dad239142ed1dfa0c60338ea2cb157/test/Utils.hs)

Edit: or alternatively to declare that if a prim module is not listed as safe then it is
unsafe thus removing the need to check that the two lists cover all existing modules.

fredrikNordvallForsberg commented:
Fixed by using the existing helper function in `Utils`, thanks! Assuming this is now okay on Travis, and you have no further suggestions, I'm tempted to merge this.

MatthewDaggitt commented on doc/user-manual/language/safe-agda.lagda.rst:
```diff
@@ -8,48 +8,49 @@
+============
 
 Standard Library
 ----------------
```
This bit about the standard library can be removed now as unsafe parts are now properly isolated.

fredrikNordvallForsberg replied:
Great, no more known issues! Fix incoming.

nad commented:
What is the current status? What has been addressed, and what has not been addressed?

fredrikNordvallForsberg commented:
> What is the current status? What has been addressed, and what has not been addressed?

On top of what is described in https://github.com/agda/agda/pull/3494#issue-243746142, I've implemented the rechecking of files if options change, added test cases and updated the user manual. I've left ``--cubical`` infective. 

I believe everything has now been addressed, in particular everything from https://github.com/agda/agda/issues/2487#issuecomment-444498567.

nad commented:
Where is the "restarting" functionality documented?

I've lost the overview over this discussion. If you think you've addressed everything, you have documented all the changes, and you have a test suite that checks that the documentation is correct (among other things), then perhaps it makes sense to merge these changes and ask for feedback.
